dev-docs/RFCs/proposals/loader-worker-triangulation.md

```diff
@@ -0,0 +1,68 @@
+# RFC: Loader Worker Triangulation
+
+* **Author**: Felix Palmer
+* **Date**: Apr, 2021
+* **Status**: **Conceptual Draft**
+
+## Summary
+
+This RFC proposes shifting the triangulation that is performed by the `earcut`
+library on solid polygons in the MVTLayer in deck.gl to a web worker,
+allowing the costly operation to be performed off the UI thread.
+
+## Motivation
+
+For layers with a large number of polygons performance suffers
+as the triangulation operation performed as part of the `updateState`
+method. As this is invoked on the main thread, the UI is unresponsive while
+this operation is performed.
+
+## Observations prior to design proposal
+
+### Cannot invoke web worker from within `updateState`
+
+Passing off computation to a web worker is an inherently async operation and thus we can't just create a worker within the implementation of the Layer component (e.g. somewhere as part of the implementation of updateState) as it would mean an async operation - which is at odds with the reactive model of Layers
+
+Thus, the invocation of the web worker to triangulate the polygon layer should happen before passing the data to the layer. This is in line with the recommendation in https://deck.gl/docs/developer-guide/performance regarding passing attribute buffer directly to Layers
+
+### Preprojection
+
+Deep within the solid polygon layer we find the call out to earcut, including an optional preprojection [layers/src/solid-polygon-layer/polygon.js]. The preprojection is invoked for LNGLAT coordinate system and thus whether it occurs is coupled to the layer and potentially viewport configuration
+
+When invoking the triangulation in the worker we need to specify whether to project or not. This is a bit messy as we are essentially duplicating the internal logic of the Layer, but I don't see a way around it. In practice most MVTLayers will likely be LNGLAT so having this as the default should mean that in practice most people will not be tripped up.
+
+### Moving earcut to loaders.gl
+
+While I was hesitant about putting the triangulation within loaders.gl, the advantage of doing the work there (rather than in a worker launched within deck.gl) is that we only have to transfer the data to-and-from the worker once. If the worker was launched from the deck.gl code it would lead to two workers being used, adding latency.
+
+### Earcut removal from deck.gl
```

Comment:
The `SolidPolygonLayer` is a general-purpose layer and should keep its triangulation capability. For the sake of this proposal it is desirable to limit the change inside the MVT use case. 

While all polygon/GeoJSON rendering could potentially benefit from web-worker-based triangulation, we do not have data regarding when the cross-thread overhead becomes justified. It would be great to run some benchmark tests on a separate track to compare the processing time on the main thread vs worker with different polygon count and vertex count per polygon.

On the other hand, MVT parsing is already performed on workers, so the cost of performing additional triangulation there is very low.

Reply:
Thanks for this input. Good to know that is what you expect of the general purpose layer. I agree that completely removing the triangulation would make the layer potentially harder to use for consumers of the library as they couldn't simply.

I've made a note of your suggestion for benchmark research and will try to come back to it